guppylang/std/qsystem/random.py

```diff
@@ -99,3 +101,27 @@ def shuffle(self: "RNG", array: array[SHUFFLE_T, SHUFFLE_N]) -> None:
             # https://github.com/CQCL/guppylang/issues/924
             if i != j:
                 mem_swap(array[i], array[j])
+
+    @guppy(qsystem_random)
+    @no_type_check
+    def discrete_rv(self: "RNG", weights: array[float, DISCRETE_N]) -> int:
```

Comment:
Note, this is a "one-shot" version of the method. If the user wishes to sample multiple times from the same distribution, it would be slightly more efficient to provide a class that they can initialize with the array and then call a "sample" method on. Then we'd only have to sanity-check the weights once, and we could cache the partial sums. On reflection, since there is little downside even if they only want one sample, I think that would be preferable and we can dispense with the one-shot version (assuming all this is possible in guppy).